File: HDOJ/1250.java
////////////////////System Comment////////////////////
////Welcome to Hangzhou Dianzi University Online Judge
////http://acm.hdu.edu.cn
//////////////////////////////////////////////////////
////Username: changmu
////Nickname: ³¤Ä¾
////Run ID: 
////Submit time: 2014-08-28 17:06:44
////Compiler: Java
//////////////////////////////////////////////////////
////Problem ID: 1250
////Problem Title: 
////Run result: Accept
////Run time:218MS
////Run memory:12112KB
//////////////////System Comment End//////////////////
import java.util.Scanner;
import java.math.BigInteger;

public class Main{
	public static void main(String[] args){
		Scanner cin = new Scanner(System.in);
		BigInteger a, b, c, d;
		BigInteger[] arr = new BigInteger[10000];
		arr[1] = arr[2] = arr[3] = arr[4] = BigInteger.ONE;
		arr[5] = BigInteger.valueOf(4);
		for(int i = 6; i < 10000; ++i)
			arr[i] = arr[i-1].add(arr[i-2]).add(arr[i-3]).add(arr[i-4]);
		int n;
		while(cin.hasNext()){
			n = cin.nextInt();
			System.out.println(arr[n]);
		}
	}
}

File: HDOJ/1250.cpp
////////////////////System Comment////////////////////
////Welcome to Hangzhou Dianzi University Online Judge
////http://acm.hdu.edu.cn
//////////////////////////////////////////////////////
////Username: changmu
////Nickname: ³¤Ä¾
////Run ID: 
////Submit time: 2014-10-20 19:06:01
////Compiler: Visual C++
//////////////////////////////////////////////////////
////Problem ID: 1250
////Problem Title: 
////Run result: Accept
////Run time:31MS
////Run memory:9240KB
//////////////////System Comment End//////////////////
#include <stdio.h>
#include <string.h>

#define mod 1000000000

int arr[10001][230];

void Add(int i, int j) {
	for(int k = 0; k < 230; ++k) {
		arr[i][k] += arr[j][k];
		if(arr[i][k] >= mod) {
			arr[i][k] -= mod;
			++arr[i][k+1];
		}
	}
}

int main() {
	int i, j, n;
	for(i = 1; i < 5; ++i)
		arr[i][0] = 1;
	for(i = 5; i < 10001; ++i) {
		for(j = i - 4; j < i; ++j)
			Add(i, j);
	}
	while(scanf("%d", &n) == 1) {
		for(j = 229; ; --j)
			if(arr[n][j]) {
				printf("%d", arr[n][j]);
				break;
			}
		for(--j; j >= 0; --j)
			printf("%09d", arr[n][j]);
		printf("\n");
	}
	return 0;
}